atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/cache/SnapshotStoreImplTest.java

```
@@ -0,0 +1,101 @@
+    private static final Sequence SEQUENCE_1 = Sequence.of(1337L);
+    private static final Sequence SEQUENCE_2 = Sequence.of(8284L);
+    private static final StartTimestamp TIMESTAMP_1 = StartTimestamp.of(42L);
+    private static final StartTimestamp TIMESTAMP_2 = StartTimestamp.of(31415925635L);
+    private static final StartTimestamp TIMESTAMP_3 = StartTimestamp.of(404L);
+    private static final StartTimestamp TIMESTAMP_4 = StartTimestamp.of(10110101L);
+    private static final ValueCacheSnapshot SNAPSHOT_1 = ValueCacheSnapshotImpl.of(HashMap.empty(), HashSet.empty());
+    private static final ValueCacheSnapshot SNAPSHOT_2 = ValueCacheSnapshotImpl.of(
+            HashMap.<CellReference, CacheEntry>empty()
+                    .put(
+                            CellReference.of(
+                                    TableReference.createFromFullyQualifiedName("t.table"),
+                                    Cell.create(new byte[] {1}, new byte[] {1})),
+                            CacheEntry.locked()),
+            HashSet.empty());
+    private SnapshotStore snapshotStore;
+
+    @Before
+    public void before() {
+        snapshotStore = new SnapshotStoreImpl();
+    }
+
+    @Test
+    public void singleSnapshotStoredForMultipleTimestamps() {
+        snapshotStore.storeSnapshot(SEQUENCE_1, TIMESTAMP_1, SNAPSHOT_1);
+        snapshotStore.storeSnapshot(SEQUENCE_1, TIMESTAMP_2, SNAPSHOT_1);
+        snapshotStore.storeSnapshot(SEQUENCE_1, TIMESTAMP_3, SNAPSHOT_1);
+        snapshotStore.storeSnapshot(SEQUENCE_2, TIMESTAMP_4, SNAPSHOT_2);
+
+        assertSnapshotsEqualForTimestamp(SNAPSHOT_1, TIMESTAMP_1, TIMESTAMP_2, TIMESTAMP_3);
+        assertSnapshotsEqualForTimestamp(SNAPSHOT_2, TIMESTAMP_4);
+    }
+
+    @Test
+    public void snapshotsNotOverwrittenForSameSequence() {
+        snapshotStore.storeSnapshot(SEQUENCE_1, TIMESTAMP_1, SNAPSHOT_1);
+        snapshotStore.storeSnapshot(SEQUENCE_1, TIMESTAMP_2, SNAPSHOT_2);
```

Should this even fail silently or throw?

---

You're right, probably should fail loudly.